provider/gce/google/errors.go

```diff
@@ -55,3 +60,57 @@ func NewMissingConfigValue(key, field string) error {
 func (err InvalidConfigValueError) Error() string {
 	return fmt.Sprintf("invalid config value (%s) for %q: %v", err.Value, err.Key, &err.Err)
 }
+
+// HandleCredentialError determines if a given error relates to an invalid credential.
+// If it is, the credential is invalidated. Original error is returned untouched.
+func HandleCredentialError(err error, ctx context.ProviderCallContext) error {
+	maybeInvalidateCredential(err, ctx)
+	return err
+}
+
+func maybeInvalidateCredential(err error, ctx context.ProviderCallContext) bool {
+	if ctx == nil {
+		return false
+	}
+	if !HasDenialStatusCode(err) {
+		return false
+	}
+
+	invalidateErr := ctx.InvalidateCredential("google cloud denied access")
+	if invalidateErr != nil {
+		logger.Warningf("could not invalidate stored google cloud credential on the controller: %v", invalidateErr)
+	}
+	return true
+}
+
+// HasDenialStatusCode determines if the given error was caused by an invalid credential, i.e. whether it contains a
+// response status code that indicates an authentication failure.
+func HasDenialStatusCode(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// http/url.Error is constructed with status code in mind and, at the time of writing for go-1.10,
+	// contains response status code and description in error.Error.
+	// We have to examine the error message to determine whether the error is related to authentication failure.
+	if cause, ok := errors.Cause(err).(*url.Error); ok {
+		for code, desc := range AuthorisationFailureStatusCodes {
+			if strings.Contains(cause.Error(), fmt.Sprintf(": %v %v", code, desc)) {
+				return true
+			}
+		}
+	}
+	return false
+
+}
+
+// AuthorisationFailureStatusCodes contains http status code nad description that signify authorisation difficulties.
+var AuthorisationFailureStatusCodes = map[int]string{
```

Comment:
I wonder if this is useful or needed in other providers? It seems quite generic.

Reply:
Not really generic at all...

For one, for example, azure only has the codes not the description. Also the code for bad request is only applicable to oauth2-compliant clients. So for example, azure does not send it for authentication issues but gce does :)

So, at this stage, fairly provider specific. If I'll encounter duplication in other providers, will pull it into provider/common area or something...